cli/snapshot.go

```diff
@@ -85,32 +94,69 @@ func (runOpts *runOptionsType) DumpSnapshot(ctx *cli.Context) error {
 	return runOpts.CopySnapshot(ctx, os.Stdout)
 }
 
-func (runOpts *runOptionsType) getRootDev(fsPath string) (*system.MountInfo, error) {
-	dataDevID, err := system.GetDeviceIDFromPath(runOpts.dataStore)
+func checkSnapshotPreconditions(
+	fsPath string,
+	dataPath string,
+) (*system.MountInfo, error) {
+	var err error
+
+	rootDevID, err := system.GetDeviceIDFromPath(fsPath)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err,
+			"error getting device id from path '%s'", fsPath)
 	}
-	rootDevID, err := system.GetDeviceIDFromPath(fsPath)
+	dataDevID, err := system.GetDeviceIDFromPath(dataPath)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err,
+			"error getting device id from path '%s'", dataPath)
 	}
 	if dataDevID == rootDevID {
-		log.Errorf(errMsgDataPartF,
-			runOpts.dataStore, runOpts.dataStore)
+		log.Errorf(errMsgDataPartFmt, dataPath, dataPath)
 		return nil, errors.Errorf(
-			"Data store (%s) is located on filesystem %s",
-			runOpts.dataStore, fsPath)
+			"data store (%s) is located on filesystem %s",
+			dataPath, fsPath)
+	}
+	rootDev, err := system.GetMountInfoFromDeviceID(rootDevID)
+	if err != nil {
+		return nil, err
+	} else if rootDev.FSType == "overlay" {
+		log.Error("overlay filesystem detected, " +
+			"filesystem type not supported")
+		log.Error("please specify the path to the " +
+			"block-device instead")
+		return nil, fmt.Errorf(
+			"snapshot: filesystem 'overlay' not supported")
 	}
 
-	return system.GetMountInfoFromDeviceID(rootDevID)
+	return rootDev, nil
 }
 
 func prepareOutStream(out io.WriteCloser, compression string) (io.WriteCloser, error) {
+
+	var pSize uint64
+
+	if f, ok := out.(*os.File); ok {
+		pSize = uint64(system.GetPipeSize(int(f.Fd())))
```

Comment:
It's just a way of detecting if the pipe is read on the other side - It's needed because this will prohibit the progress bar from cluttering possible prompts from other process(es).

Reply:
I create a start condition for the progress bar print function based on the pipe (or actually input) buffer size.

Reply:
Alright, that makes sense. Please add some comments about that though. The code doesn't make it clear why this is needed.